### DISCUSSION TITLE
cweb for papers?

### DISCUSSION THREAD

**Discussion by zmajeed** [shreevatsa/webWEB#10] [General]
https://youtu.be/vznwJU-WdUo
https://desfontain.es/privacy/latex-to-html.html

got me thinking about cweb for papers - there's a heartening trend of providing source code in or with papers - I wonder if there are any examples of papers written in cweb?

notebooks are another step in the evolution of papers and technical presentations - could cweb be enhanced to support something like interactive documentation - this is a relatively new idea of documentation as a sandbox?

**Comment by texdraft** [comment]
Semi-related (regarding notebooks): Something I call &ldquo;incremental Literate Programming&rdquo; has enjoyed success in educational literature for a while. Recent examples are &ldquo;[Kilo](https://viewsourcecode.org/snaptoken/kilo/)&rdquo; and [*Crafting Interpreters*](https://craftinginterpreters.com/). The idea is that sections can be superseded, not just added on to. The earliest instance of this technique I am aware of is Kernighan and Plaugher's [*Software Tools in Pascal*](https://seriouscomputerist.atariverse.com/media/pdf/book/Software%20Tools%20in%20Pascal.pdf), written in 1981 (nearly contemporaneous with `WEB`&mdash;see also @shreevatsa's page [here](https://shreevatsa.github.io/tex/program/pooltype/#other-problems-with-pascal)). Note that the incremental model is actually more general than the concatenative model.

Adding incremental abilities to `CWEB` might be feasible (repurpose the change file notation, perhaps?). I think the combination of named sections of code (which Kilo and *Crafting Interpreters* lack) and being able to update the code progressively would create quite a powerful documentation engine.

**Reply by shreevatsa** [reply]
Like the two examples above (Kilo and Crafting Interpreters), another presentation that I thought was nice:

- [Git Internals - Learn By Building Your Own Git: μgit (ugit)](https://www.leshenko.net/p/ugit) by Nikita Leshenko

(And do let me know if you're aware of more such examples…)

And (this is even less related, but I just remembered…), there's an intriguing comment by Knuth in his letter to Dijkstra [A review of "Structured Programming"](http://i.stanford.edu/pub/cstr/reports/cs/tr/73/371/CS-TR-73-371.pdf) (1972):

> Herein lies the dilemma, and the conflict. A notation which expresses the separation of tasks nicely eventually gets into difficulties when there are many loose threads running through a single pearl. Yet the pearls are valuable as a means of coordinating the individual design decisions.

> Perhaps the new display terminals will provide the answer, as a notation for programs that will provide an appropriate reflection of the structure as we might have it in our minds. You have perhaps had a dream much like mine: Wouldn't it be nice to have a glorious system of complete maps of the world, whereby one could (by turning dials) increase or decrease the scale at will? A similar thing can be achieved for programs, rather easily, when we give the programs a hierarchic structure like those constructed step-wise. It is not hard to imagine a computing system which displays a program in such a way that, by pressing an-appropriate button, one can replace the name of a routine by its expansion on the next level, or conversely.

> The independent design decisions (i.e., the pearls) could be identified in such a system by marginal classification codes, saying which pearl each line belongs to. An example appears below. This seems to me to be a reasonable way to bring the appropriate context into each pearl, yet retaining the pearl's identity. (We need another word for pearl, though; what should it be?)

(Search for "pearl" in the PDF file for the example program, twice—PDF pages 9 of 26, and 23 of 26.) (I haven't actually read and understood what's going on in those examples, but there seems to be some connection with the later idea of literate programming.)

**Reply by texdraft** [reply]
@shreevatsa Here's something pretty cool: https://observablehq.com/@bobkerns/maclisp

It's like a super-meta semi-literate programming system. You can edit the prose and the code, and everything's outlined and hyperlinked together. But it's not really incremental.

**Comment by zmajeed** [comment]
This implies section versioning - versions could be tied to git commits - and commit prefixes could be used as a control code argument

yeah - an enhanced change control code is a possibility

**Reply by texdraft** [reply]
Some sort of integrated version control + `CWEB` system might be interesting. Imagine if the woven documentation for TeX allowed you to see the history of any given section (as given in [`tex82.bug`](https://ctan.org/tex-archive/systems/knuth/dist/errata)).

**Reply by shreevatsa** [reply]
That would be really cool; just imagining literate programs where each commit can itself be like a literate program annotating the diffs it introduces (hey we already write commit messages, and code review comments on individual lines of code), and you have a slider or something where you can go back in time and see older versions of a program or particular section. :-)

**Comment by shreevatsa** [comment]
I imagine there may not be too many papers written in CWEB, at least recent ones, but in some sense each CWEB program is itself like a paper. And Knuth's WEB programs like the two he wrote for Bentley's column can be considered "papers" as (they appeared in the *Communications of the ACM*, and) each of them was introducing a somewhat new idea, in addition to introducing literate programming.

One issue with both of them, and with regarding each WEB/CWEB program as a paper, is that for a paper one probably wants to highlight specific parts of the code, while with WEB/CWEB the entire program is presented, even the "boring" parts. This was pointed out by Harold Thimbleby, who wrote a system called `cweb` based on WEB, before the CWEB that we're talking about existed. His papers are here: http://www.harold.thimbleby.net/cv/warp.html and includes "Explaining Code for Publication" from as recent as 2003, and even a 2018 paper that I'm only seeing just now, about a tool for "reverse literate programming" (extracting several `program.c` files from `paper.tex`, rather than CWEB's `program.w` → {`program.c`, `program.tex`}): https://github.com/haroldthimbleby/relit
